sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java

```diff
@@ -189,6 +533,725 @@ private Write() {}
+     */
+    static final class PartitionQueryResponseToRunQueryRequest
+        extends DoFn<PartitionQueryPair, RunQueryRequest> {
+
+      /**
+       * When fetching cursors that span multiple pages it is expected (per <a
+       * href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.PartitionQueryRequest">
+       * PartitionQueryRequest.page_token</a>) for the client to sort the cursors before processing
+       * them to define the sub-queries. So here we're defining a Comparator which will sort Cursors
+       * by the first reference value present, then comparing the reference values
+       * lexicographically.
+       */
+      static final Comparator<Cursor> CURSOR_REFERENCE_VALUE_COMPARATOR;
+
+      static {
+        Function<Cursor, Optional<Value>> firstReferenceValue =
+            (Cursor c) ->
+                c.getValuesList().stream()
+                    .filter(
+                        v -> {
+                          String referenceValue = v.getReferenceValue();
+                          return referenceValue != null && !referenceValue.isEmpty();
+                        })
+                    .findFirst();
+        Function<String, String[]> stringToPath = (String s) -> s.split("/");
+        // compare references by their path segments rather than as a whole string to ensure
+        // per path segment comparison is taken into account.
+        Comparator<String[]> pathWiseCompare =
+            (String[] path1, String[] path2) -> {
+              int minLength = Math.min(path1.length, path2.length);
+              for (int i = 0; i < minLength; i++) {
+                String pathSegment1 = path1[i];
+                String pathSegment2 = path2[i];
+                int compare = pathSegment1.compareTo(pathSegment2);
+                if (compare != 0) {
+                  return compare;
+                }
+              }
+              if (path1.length == path2.length) {
+                return 0;
+              } else if (minLength == path1.length) {
+                return -1;
+              } else {
+                return 1;
+              }
+            };
+
+        // Sort those cursors which have no firstReferenceValue at the bottom of the list
+        CURSOR_REFERENCE_VALUE_COMPARATOR =
+            Comparator.comparing(
+                firstReferenceValue,
+                (o1, o2) -> {
+                  if (o1.isPresent() && o2.isPresent()) {
+                    return pathWiseCompare.compare(
+                        stringToPath.apply(o1.get().getReferenceValue()),
+                        stringToPath.apply(o2.get().getReferenceValue()));
+                  } else if (o1.isPresent()) {
+                    return -1;
+                  } else {
+                    return 1;
+                  }
+                });
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) {
+        PartitionQueryPair pair = c.element();
```

Comment:
It seems this logic isn't tested, just the comparator itself. Can you add one that calls `runFunction`?

Reply:
Unit test added in addition to existing integration test.